src/aspire/source/coordinates.py

```diff
@@ -0,0 +1,493 @@
+        corner of the image.
+        :param data: A 2D numpy array representing a micrograph
+        :param coord: A list of integers: (lower left X, lower left Y, X, Y)
+        """
+        start_x, start_y, size_x, size_y = coord
+        return data[start_y : start_y + size_y, start_x : start_x + size_x]
+
+    def _images(self, start=0, num=np.inf, indices=None):
+        """
+        Given a range or selection of indices, returns an Image stack
+        of the particles specified. Note that the indices refer to the order
+        of the particles loaded in this *specific* CoordinateSource. This may
+        not correspond to the particles in the original source on disk, if some
+        particles were excluded due to their box not fitting into the mrc
+        dimensions. Thus, the exact particles returned are a function of the
+        `particle_size`.
+        :param start: Starting index (default: 0)
+        :param num: number of images to return starting from `start` (default: numpy.inf)
+        :param indices: A numpy array of integer indices
+        """
+        if indices is None:
+            indices = np.arange(start, min(start + num, self.n))
+
+        logger.info(f"Loading {len(indices)} images from micrographs")
+
+        selected_particles = [self.particles[i] for i in indices]
+        # initialize empty array to hold particle stack
+        im = np.empty(
+            (len(indices), self._original_resolution, self._original_resolution),
+            dtype=self.dtype,
+        )
+
+        # group particles by micrograph in order to
+        # only open each one once
+        grouped = defaultdict(list)
+        # this creates a dict of the form
+        # { mrc_index : list of coords in that mrc, with order preserved }
+        for mrc_index, coord in selected_particles:
+            grouped[mrc_index].append(coord)
+
+        for mrc_index, coord_list in grouped.items():
+            # get explicit filepath from cached list
+            fp = self.mrc_paths[mrc_index]
+            with mrcfile.open(fp) as mrc_in:
+                arr = mrc_in.data.astype(self.dtype)
+            if arr.shape != self.mrc_shape:
+                raise ValueError(
+                    f"Shape of {fp} is {arr.shape}, but expected {self.mrc_shape}"
+                )
+            # create iterable of the coordinates in this mrc
+            # we don't need to worry about exhausting this iter
+            # because we know it contains the exact number of particles
+            # selected from this micrograph
+            coord = iter(coord_list)
+            # iterate through selected particles
+            for i, particle in enumerate(selected_particles):
+                idx = particle[0]
+                # we stop and populate the image stack every time
+                # we hit a particle whose location is this micrograph
+                if idx == mrc_index:
```

Comment:
I don't believe it's inefficient as a function of the number of MRC files. The optimization for MRC files has already been done in the `ddict` step. 

The following step only happens once per micrograph:

```python
for mrc_index, coord_list in grouped.items():
            # get explicit filepath from cached list                                                                                                                 
            fp = self.mrc_paths[mrc_index]
            with mrcfile.open(fp) as mrc_in:
                arr = mrc_in.data.astype(self.dtype)
```

The second loop is necessary because we need to populate the return `Image` stack with the particles *in the order they were requested*. The cost of looping through lists of indices is very small compared to the cost of opening the .mrc file. This step also only happens once per micrograph, with all particles from that micrograph being populated in one pass.

So even if we got the original indices directly from the `ddict`, we'd be populating the `Image` stack out of order anyway, which would involve searching through `selected_particles` for the indices, with a similar cost to using this method.